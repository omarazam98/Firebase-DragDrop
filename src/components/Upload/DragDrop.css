#dropGuide{
    border: dashed grey 4px;
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
}

#dropHelpLabel {
    textAlign: center;
    color: grey;
    fontSize: 36;
}

#dropZone {
    display: inline-block;
    position: relative;
}