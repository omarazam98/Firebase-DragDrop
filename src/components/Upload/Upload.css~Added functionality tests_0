.form {
    display: flex;
    justify-content: center;
    align-items: center;
    flex-direction: column;
    padding: 1rem;
    margin: 1rem;
    background: lightseagreen;
}

.error {
    font-size: 14px;
    text-align: center;
    color: red;
}

input[type='file'] {
    color: transparent;
}
