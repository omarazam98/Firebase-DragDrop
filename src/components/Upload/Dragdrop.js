import * as React from 'react';
import './DragDrop.css';

interface DragdropState {
    dragging: boolean; // state that keeps track of if cursor is over dragdrop area
    dragCounter: number; // used to control dragging state and css of drop div's
};

const INITIAL_STATE: DragdropState = {
    dragging: false,
    dragCounter: 0,
};

export class Dragdrop extends React.Component {

    constructor(props) {
        super(props);
        this.state = INITIAL_STATE;
        this.dropRef = React.createRef() // ref created for drop div
        this.handleDrag = this.handleDrag.bind(this);
        this.handleDragIn = this.handleDragIn.bind(this);
        this.handleDragOut = this.handleDragOut.bind(this);
        this.handleDrop = this.handleDrop.bind(this);
    }

    handleDrag(e) {
        e.preventDefault(); // these 2 lines of coded are found in each handler function, stops browser from its default action of opening a file
        e.stopPropagation()
    };

    handleDragIn(e) {
        e.preventDefault();
        e.stopPropagation();
        this.setState(function (prevState, props) {
            return {dragCounter: (prevState.dragCounter + 1)} // when cursor on drag area the counter == 1, this changes the state
        });
        if (e.dataTransfer.items && e.dataTransfer.items.length > 0) { // Only changes the state of dragging if user is dragging a file
            this.setState(function (prevState, props) {
                return {dragging: true}
            });
        }
    };

    handleDragOut(e) {
        e.preventDefault();
        e.stopPropagation();
        this.setState(function (prevState, props) {
            return {dragCounter: (prevState.dragCounter - 1)} // when cursor on drag area the counter == 1, this changes the state
        }, () =>{
            if (this.state.dragCounter === 0) { // cursor leaves, counter == 0 dragging is set to false
                this.setState(function (prevState, props) {
                    return {dragging: false}
                });
            }
        });

    };

    handleDrop(e) {
        e.preventDefault();
        e.stopPropagation();
        this.setState(function (prevState, props) {
            return {dragging: false} // when file is dropped dragging is done
        });
        this.props.handleDrop(e.dataTransfer.files) // passes files to handleDrop method in props that can be accessed by this components parent
    };

    componentDidMount() {
        let div = this.dropRef.current;
        div.addEventListener('dragenter', this.handleDragIn); // add event listeners to core events found in div docs
        div.addEventListener('dragleave', this.handleDragOut);
        div.addEventListener('dragover', this.handleDrag);
        div.addEventListener('drop', this.handleDrop);
    }

    componentWillUnmount() {
        let div = this.dropRef.current
        div.removeEventListener('dragenter', this.handleDragIn);
        div.removeEventListener('dragleave', this.handleDragOut);
        div.removeEventListener('dragover', this.handleDrag);
        div.removeEventListener('drop', this.handleDrop)
    }

    render() {
        return (
            <div id = 'dropZone' ref={this.dropRef}>
                {this.state.dragging && // conditional rendering (when state is in dragging)
                <div id="dropGuide">
                    <div id="dropHelpLabel">
                        DROP HERE
                    </div>
                </div>
                }
                {this.props.children}
            </div>
        )
    }
}

export default Dragdrop;
