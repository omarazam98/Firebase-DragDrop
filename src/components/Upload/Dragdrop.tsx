import * as React from 'react';
import './DragDrop.css';

interface DragdropState {
  draggingFile: boolean;
  dragging: boolean;
}

const INITIAL_STATE: DragdropState = {
  draggingFile: false,
  dragging: false,
};

class Dragdrop extends React.Component<any, DragdropState> {
  constructor(props) {
    super(props);
    this.state = INITIAL_STATE;
    this.handleDrag = this.handleDrag.bind(this);
    this.handleDragEnter = this.handleDragEnter.bind(this);
    this.handleDragExit = this.handleDragExit.bind(this);
    this.handleDrop = this.handleDrop.bind(this);
  }

  handleDrag(e) {
    // stops browser from its default action of opening a file
    e.preventDefault();
    e.stopPropagation();
  }

  handleDragEnter(e) {
    e.preventDefault();
    e.stopPropagation();
    this.setState((prevState, props) => {
      return { dragging: true };
    });
    // Only change the state of dragging if user is dragging a file
    if (e.dataTransfer.items && e.dataTransfer.items.length > 0) {
      this.setState((prevState, props) => {
        return { draggingFile: true };
      });
    }
  }

  handleDragExit(e) {
    e.preventDefault();
    e.stopPropagation();
    this.setState(
      (prevState, props) => {
        return { dragging: false };
      },
      () => {
        if (!this.state.dragging) {
          this.setState((prevState, props) => ({ draggingFile: false }));
        }
      });
  }

  handleDrop(e) {
    e.preventDefault();
    e.stopPropagation();
    // when file is dropped dragging is done
    this.setState(() => {
      return {
        draggingFile: false,
      };
    });
    // passes files to handleDrop method in props
    this.props.handleDrop(e.dataTransfer.files);
  }

  componentDidMount() {
    const div = document.getElementById('dropZone');
    if (div) {
      div.addEventListener('dragenter', this.handleDragEnter);
      div.addEventListener('DragExit', this.handleDragExit);
      div.addEventListener('dragover', this.handleDrag);
      div.addEventListener('drop', this.handleDrop);
    }
  }

  componentWillUnmount() {
    const div = document.getElementById('dropZone');
    if (div) {
      div.removeEventListener('dragenter', this.handleDragEnter);
      div.removeEventListener('DragExit', this.handleDragExit);
      div.removeEventListener('dragover', this.handleDrag);
      div.removeEventListener('drop', this.handleDrop);
    }
  }

  render() {
    return (
      <div id = "dropZone">
        {this.state.draggingFile &&
        <div id="dropGuide">
            <div id="dropHelpLabel">
                DROP HERE
            </div>
            <div>

            </div>
        </div>
        }
        {this.props.children}
      </div>
    );
  }
}

export default Dragdrop;
