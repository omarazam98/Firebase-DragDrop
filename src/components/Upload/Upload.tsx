import * as React from 'react';
import Firebase, {withFirebase} from '../Firebase';
import {ChangeEvent} from "react";
import * as firebase from 'firebase'
import Dragdrop from "./Dragdrop";
import './Upload.css';


class UploadBase extends React.Component<any, any> {
    docPathDB: String;
    docPathStorage: String;

    constructor(props: Firebase) {
        super(props);
        this.docPathDB = "Files/documentTest/";
        this.docPathStorage = "test/documents/";
        this.state = {
            userName: 'userX',
            uploading: false,
            uploaded: false,
            percent: 0,
            file: null,
            error: null
        }
    }
    /**
     * Gets file from browse button and sets file state
     * @param {ChangeEvent<HTMLInputElement>} e : event from file browse button
     */
    handleFileSelect = (e: ChangeEvent<HTMLInputElement>) => {
        this.setState({uploaded: false})
        this.setState({file: e.target.files![0]})

    }
    /**
     * Gets file from Dragdrop component and sets file state
     * @param {FileList} files : files from Dragdrop
     */
    handleDropSelect = (files: FileList) => {
        this.setState({uploaded: false})
        this.setState({file: files[0]})
    }

    /**
     * Uploads file from file state and sends to firebase
     * stored in cloud storage at this.docPathStorage + filename
     * stored in firestore database as URL to storage under username (default userX)
     * also updates a progress bar as file is uploaded
     */
    uploadFile = () => {
        if(this.state.file) {
            const file = this.state.file;
            this.setState({uploading: true});
            try {
                const storageRef = this.props.firebase.storage.ref(this.docPathStorage + file.name); //Creates a storage reference at filepath
                const task = storageRef.put(file); //puts file in that reference
                task.on('state_changed',
                    function progress(snapshot: firebase.storage.UploadTaskSnapshot) {
                        const uploader = document.getElementById("uploader") as HTMLProgressElement;
                        const percentage = (snapshot.bytesTransferred / snapshot.totalBytes) * 100;
                        uploader.value = percentage;
                    }, function error(err: Error) {
                        console.log(err);
                    }, () => {
                    console.log(this.state.uploading)
                        this.setState({uploaded: true});
                        this.setState({uploading: false});
                        console.log("Uploaded file to " + 'users/' + this.state.userName + '/' + this.docPathDB);
                        task.snapshot.ref.getDownloadURL().then((downloadURL: String) => {
                            this.props.firebase.db.doc('users/' + this.state.userName + '/' + this.docPathDB).set(
                                {
                                    URL: downloadURL
                                }
                            );
                        });
                    });
            }catch(err){
                this.setState({error:err});
                this.setState({uploading: false});
            }
        }

    }

    render() {
        return (
            <div>
                <h1>Uploading Page</h1>
                <p>This page is to test uploading</p>

                <Dragdrop handleDrop={this.handleDropSelect}>
                <div className='container'>
                    <div className='form'>
                        <div>
                            <label htmlFor="fileButton" className="btn">Choose file or drop here</label>
                            <input type="file" id="fileButton" onChange={this.handleFileSelect}/>
                        </div>
                        <div>
                            {this.state.uploading && <progress value="0" max="100" id="uploader">0%</progress>}
                            {this.state.uploaded && <p id='checkmark' style={{display: "inline"}}>&#9989;</p>}
                        </div>
                        {this.state.file && <p>{this.state.file.name}</p>}
                        {this.state.file && <button id ="uploadButton" onClick={this.uploadFile}>Upload</button>}
                    </div>
                    <code>
                        {this.state.error ? <span className='error'>{this.state.error.message}</span> : null}
                    </code>
                </div>
            </Dragdrop>
            </div>
        );
    }
}

const Upload = withFirebase(UploadBase);
export {Upload, UploadBase};
