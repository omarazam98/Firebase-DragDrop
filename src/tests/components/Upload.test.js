import React from 'react';
import { Upload } from '../../components/Upload/Upload';
import { shallow } from 'enzyme'


let wrapper;

beforeEach(() => {
    wrapper = shallow(
        <Upload/>);
});

test('has a valid snapshot', () => {
    expect(wrapper).toMatchSnapshot();
});

it('Select file changes state', () => {
    const file = new File(["Hello World"], "file.txt");
    expect(wrapper.state('file')).toBeDefined();
    const mockEvent = {
        target: {
            files: [
                file
            ]
        }};
    expect(wrapper.state.file).toBeUndefined();
});

it('Clicking upload button calls uploadFile', () => {
    const f = new File(["Hello World"], "file.txt");
    const spy = jest.spyOn(wrapper.instance(), 'uploadFile');
    wrapper.setState({file: f});
    wrapper.find('#uploadButton').simulate('click');
    expect(spy).toHaveBeenCalled();
});