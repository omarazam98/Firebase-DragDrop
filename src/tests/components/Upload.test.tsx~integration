import React from 'react';
import { Upload } from '../../components/Upload/Upload';
import { shallow } from 'enzyme'


let wrapper;
let mockEvent;

beforeEach(() => {
    wrapper = shallow(
        <Upload/>);
});

it('has a valid snapshot', () => {
    expect(wrapper).toMatchSnapshot();
});

it('File not added to state if invalid type', () => {
    const file = new File(["Hello World"], "file.txt");
    mockEvent = {
        target: {
            files: [
                file
            ]
        }};
    wrapper.instance().handleFileSelect(mockEvent);
    expect(wrapper.state('file')).toBeNull();
    expect(wrapper.state('error').message).toBe('File type not accepted')
});

it('Clicking upload button calls uploadFile', () => {
    const f = new File(["Hello World"], "file.txt");
    const spy = jest.spyOn(wrapper.instance(), 'uploadFile');
    wrapper.setState({file: f});
    wrapper.find('#uploadButton').simulate('click');
    expect(spy).toHaveBeenCalled();
});