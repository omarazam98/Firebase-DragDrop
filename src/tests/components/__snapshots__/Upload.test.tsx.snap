// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has a valid snapshot 1`] = `
<div>
  <Dragdrop
    handleDrop={[Function]}
  >
    <div
      className="form"
    >
      <div>
        <label
          className="btn"
          htmlFor="fileButton"
        >
          Choose file or drop here
        </label>
        <input
          id="fileButton"
          onChange={[Function]}
          type="file"
        />
      </div>
      <div />
    </div>
    <code />
  </Dragdrop>
</div>
`;
